torch/_inductor/codegen/cpp.py

```diff
@@ -835,6 +835,14 @@ def index_expr(expr, dtype):
 class CppOverrides(OpOverrides):
     """Map element-wise ops to C++"""
 
+    @staticmethod
+    def add(a, b):
+        return f"decltype({a})({a} + {b})"
```

Comment:
Won't this be wrong for:
int + float
?

Maybe we need to do something with C++ templates to override default type promotion behavior.

Reply:
If this is wrong we should also revisit `mul` a couple lines below this one

Reply:
> Won't this be wrong for:
> int + float

Does the upper-level lowering make sure the explicit `to_dtype` is inserted into the IR and mixed data types won't happen in this level? @jansel Any specific case in your mind that could trigger this? I can see how to fix then.

Reply:
Hrm, maybe this will be fine.  I worry we are relying on type promotion somewhere.

Reply:
yep, I guess the `transform_args` called by `register_lowering` from inductor lowering should take care of the type promotions to match the data types of input args. I can add a compilation time assertion in the follow-up PR to guarantee the data types for these ops are the same.